File: backend/feature_extraction.py
import re
from urllib.parse import urlparse
import tldextract
import requests
from bs4 import BeautifulSoup
import socket

def extract_features(url):
    features = {}
    
    # Clean URL format
    if not re.match(r"^https?", url):
        url = "http://" + url
        
    try:
        parsed_url = urlparse(url)
        extracted_domain = tldextract.extract(url)
    except:
        return None

    domain = f"{extracted_domain.domain}.{extracted_domain.suffix}"
    
    # 1. URL Length (Phishing usually has long URLs)
    features['url_length'] = len(url)
    features['is_long_url'] = 1 if len(url) > 54 else 0
    
    # 2. Presence of @ Symbol
    features['has_at_symbol'] = 1 if '@' in url else 0
    
    # 3. IP Address in URL
    ip_pattern = re.compile(
        r'(([01]?\d\d?|2[0-4]\d|25[0-5])\.([01]?\d\d?|2[0-4]\d|25[0-5])\.'
        r'([01]?\d\d?|2[0-4]\d|25[0-5])\.([01]?\d\d?|2[0-4]\d|25[0-5])\b)|'
        r'((0x[0-9a-fA-F]{1,2})\.(0x[0-9a-fA-F]{1,2})\.(0x[0-9a-fA-F]{1,2})\.(0x[0-9a-fA-F]{1,2})\b)'
        r'(?:[a-fA-F0-9]{1,4}:){7}[a-fA-F0-9]{1,4}'
    )
    features['has_ip_address'] = 1 if ip_pattern.search(domain) else 0
    
    # 4. Number of hyphens in domain
    features['num_hyphens'] = domain.count('-')
    features['has_hyphen_in_domain'] = 1 if '-' in domain else 0
    
    # 5. Number of subdomains
    subdomain = extracted_domain.subdomain
    num_dots = subdomain.count('.') if subdomain else 0
    features['num_subdomains'] = num_dots + 1 if subdomain else 0
    
    # 6. HTTP vs HTTPS
    features['is_https'] = 1 if parsed_url.scheme == 'https' else 0
    
    # 7. Suspicious Words in URL
    suspicious_words = ['secure', 'account', 'webscr', 'login', 'ebayisapi', 'signin', 'banking', 'confirm']
    features['has_suspicious_words'] = 1 if any(word in url.lower() for word in suspicious_words) else 0

    # Page Content Features
    features['has_form_with_action'] = 0
    features['has_hidden_elements'] = 0
    features['has_password_field'] = 0
    
    try:
        # Increase timeout slightly, fake user-agent to bypass basic blocks
        headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'}
        response = requests.get(url, timeout=3, headers=headers)
        
        if response.status_code == 200:
            soup = BeautifulSoup(response.text, 'html.parser')
            
            # Form actions empty or pointing elsewhere
            forms = soup.find_all('form')
            for form in forms:
                action = form.get('action', '')
                if not action or '#' in action or action.startswith('mailto:') or action.startswith('http'):
                    features['has_form_with_action'] = 1
                    break
                    
            # Check for password fields
            if soup.find('input', type='password'):
                features['has_password_field'] = 1
                
            # Check hidden elements (often used to obscure phishing templates)
            hidden_elements = soup.find_all(style=re.compile(r'display:\s*none|visibility:\s*hidden', re.I))
            if len(hidden_elements) > 2:
                features['has_hidden_elements'] = 1
                
    except Exception as e:
        # Request failed (common for bad or unreachable URLs)
        pass

    return features

File: frontend/src/featureExtraction.js
/**
 * featureExtraction.js
 * Ports backend/feature_extraction.py to pure JavaScript.
 * Extracts URL-based features only (no HTTP requests to external sites).
 * 100% client-side — works on GitHub Pages with no backend.
 */

/**
 * Extract phishing-detection features from a URL string.
 * Returns an object matching the feature columns used by the ML model:
 *   url_length, is_long_url, has_at_symbol, has_ip_address,
 *   num_hyphens, has_hyphen_in_domain, num_subdomains,
 *   is_https, has_suspicious_words,
 *   has_form_with_action, has_password_field, has_hidden_elements
 */
export function extractFeatures(rawUrl) {
  let url = rawUrl.trim();

  // Add scheme if missing
  if (!/^https?:\/\//i.test(url)) {
    url = 'http://' + url;
  }

  let parsed;
  try {
    parsed = new URL(url);
  } catch {
    return null; // Invalid URL
  }

  const hostname = parsed.hostname.toLowerCase(); // e.g. "sub.example.com"

  // --- 1. URL Length ---
  const urlLength = url.length;
  const isLongUrl = urlLength > 54 ? 1 : 0;

  // --- 2. @ Symbol ---
  const hasAtSymbol = url.includes('@') ? 1 : 0;

  // --- 3. IP Address in hostname ---
  const ipPattern = /^(\d{1,3}\.){3}\d{1,3}$/;
  const hasIpAddress = ipPattern.test(hostname) ? 1 : 0;

  // --- 4. Hyphens in domain ---
  // Extract registrable domain (last two parts): e.g. "example.com"
  const parts = hostname.split('.');
  const registrableDomain = parts.length >= 2 ? parts.slice(-2).join('.') : hostname;
  const numHyphens = (registrableDomain.match(/-/g) || []).length;
  const hasHyphenInDomain = numHyphens > 0 ? 1 : 0;

  // --- 5. Subdomains ---
  // Subdomains = everything before the registrable domain
  const subdomainPart = parts.slice(0, -2).join('.');
  let numSubdomains = 0;
  if (subdomainPart && subdomainPart !== 'www') {
    numSubdomains = subdomainPart.split('.').filter(Boolean).length;
  }

  // --- 6. HTTPS ---
  const isHttps = parsed.protocol === 'https:' ? 1 : 0;

  // --- 7. Suspicious Words ---
  const suspiciousWords = [
    'secure', 'account', 'webscr', 'login', 'ebayisapi',
    'signin', 'banking', 'confirm', 'update', 'verify',
    'paypal', 'password', 'credential', 'wallet', 'support',
  ];
  const urlLower = url.toLowerCase();
  const hasSuspiciousWords = suspiciousWords.some(w => urlLower.includes(w)) ? 1 : 0;

  // --- 8. Page content features ---
  // We cannot fetch page content client-side (CORS).
  // Heuristic: infer from URL patterns only.
  const hasFormWithAction = hasSuspiciousWords || hasIpAddress ? 1 : 0;
  const hasPasswordField = (hasSuspiciousWords && !isHttps) || hasIpAddress ? 1 : 0;
  const hasHiddenElements = hasIpAddress || (numHyphens > 2) ? 1 : 0;

  return {
    url_length: urlLength,
    is_long_url: isLongUrl,
    has_at_symbol: hasAtSymbol,
    has_ip_address: hasIpAddress,
    num_hyphens: numHyphens,
    has_hyphen_in_domain: hasHyphenInDomain,
    num_subdomains: numSubdomains,
    is_https: isHttps,
    has_suspicious_words: hasSuspiciousWords,
    has_form_with_action: hasFormWithAction,
    has_password_field: hasPasswordField,
    has_hidden_elements: hasHiddenElements,
  };
}
